tool/tsh/aws.go

```diff
@@ -17,185 +17,314 @@ limitations under the License.
-	// Teleport AWS Signing APP will resolve aws-service and aws-region to the proper Amazon API URL.
-	generatedAWSCred, err := genAndSetAWSCredentials()
-	if err != nil {
+	return awsApp.RunCommand(awsCLIBinaryName, cf.AWSCommandArgs...)
+}
+
+// awsApp is an AWS app that can start local proxies to serve AWS APIs.
+type awsApp struct {
+	cf      *CLIConf
+	profile *client.ProfileStatus
+	appName string
+
+	localALPNProxy    *alpnproxy.LocalProxy
+	localForwardProxy *alpnproxy.ForwardProxy
+	credentials       *credentials.Credentials
+	credentialsOnce   sync.Once
+}
+
+// newAWSApp creates a new AWS app.
+func newAWSApp(cf *CLIConf, profile *client.ProfileStatus, appName string) (*awsApp, error) {
+	return &awsApp{
+		cf:      cf,
+		profile: profile,
+		appName: appName,
+	}, nil
+}
+
+// StartLocalProxies sets up local proxies for serving AWS clients.
+//
+// There are two ways clients can connect to the local proxies.
+//
+// 1. client can send AWS requests to our local forward proxy by configuring
+// HTTPS_PROXY (or equivalent). The API flow looks like this:
+// clients -> local forward proxy -> local ALPN proxy -> remote server
+//
+// 2. client can send AWS requests to our local ALPN proxy directly by
+// configuring AWS endpoint URLs. The API flow looks like this.
+// clients -> local ALPN proxy -> remote server
+//
+// The first method is always preferred as the original hostname is preserved
+// through forward proxy.
+func (a *awsApp) StartLocalProxies() error {
+	if err := a.startLocalALPNProxy(); err != nil {
 		return trace.Wrap(err)
 	}
-
-	tc, err := makeClient(cf, false)
-	if err != nil {
+	if err := a.startLocalForwardProxy(); err != nil {
 		return trace.Wrap(err)
 	}
+	return nil
+}
 
-	lp, err := createLocalAWSCLIProxy(cf, tc, generatedAWSCred, tmpCert.getCert())
-	if err != nil {
-		return trace.Wrap(err)
+// close makes all necessary close calls.
+func (a *awsApp) Close() error {
+	var errs []error
+	if a.localALPNProxy != nil {
+		errs = append(errs, a.localALPNProxy.Close())
+	}
+	if a.localForwardProxy != nil {
+		errs = append(errs, a.localForwardProxy.Close())
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// GetAWSCredentials generates fake AWS credentials that are used for
+// signing an AWS request during AWS API calls and verified on local AWS proxy
+// side.
+func (a *awsApp) GetAWSCredentials() *credentials.Credentials {
+	// There is no specific format or value required for access key and secret,
+	// as long as the AWS clients and the local proxy are using the same
+	// credentials. The only constraint is the access key must have a length
+	// between 16 and 128. Here access key and secert are generated based on
+	// current profile and app name so the same values can be recreated.
+	//
+	// https://docs.aws.amazon.com/STS/latest/APIReference/API_Credentials.html
+	a.credentialsOnce.Do(func() {
+		hashData := []byte(fmt.Sprintf("%v-%v-%v", a.profile.Name, a.profile.Username, a.appName))
```

Comment:
If hashData is a secret generating it based on a user profile name, username and app name combination is not best security practise. Why we can used uuid here ?

Reply:
It's a good question. I didn't use `uuid` to avoid users changing credentials in their applications every `tsh` run. I can generate `uuid` then save them in a credentials file for persistence (i initially had one but later removed for simplicity,  but i can add it back). WDYT?

Reply:
Yes, Lets use the content of `profile.KeyPath()` to have a constant value.